sql/wsrep_sst.cc

```diff
@@ -27,6 +27,73 @@
 #include "wsrep_xid.h"
 #include <cstdio>
 #include <cstdlib>
+#include <functional>
+#include <sstream>
+#include <string>
+
+/* We don't need PSI instrumentation for regex mutex, however
+   Regex class has its constructor compiled conditionally.
+   Disable instrumentation. 
+   Regex class uses c++11 nullptr. As we are in c++03 world
+   let's define nullptr to satisfy compiler */
+#ifdef HAVE_PSI_INTERFACE
+#undef HAVE_PSI_INTERFACE
+#endif
+#define nullptr 0
+#include <sql_regex.h>
+
+/* Check if the string is valid. In method name we allow only:
+   alpha-num
+   underscore
+   dash
+   but w do not allow leading and trailing dash.
+   Method names have to be separated by colons (spaces before/after colon are allowed).
+   Below regex may seem difficult, but in fact it is not:
+
+   1. Any number of leading spaces
+  
+     2. Followed by  alpha-num character or underscore
+       3. Followed by any number of underscore/dash
+       4. Followed by alpha-num character or underscore
+       5. 3 - 4 group is optional
+     6. Followed by colon (possibly with leading and/or trailing spaces)
+     7. 2 - 6 group is optional
+
+   8. Followed by alpha-num character or underscore
+     9. Followed by any number of underscore/dash
+     10. Followed by alpha-num character or underscore
+     11. 9 -10 group is optional
+
+   11. Followed by any number of trailing spaces
+
+   It is the same as:
+   ^\s*(\w([_-]*\w)*(\s*,\s*))*\w([_-]*\w)*\s*$    */
+
+#define COLON_REGEX "([[:blank:]]*,[[:blank:]]*)"
+#define METHOD_REGEX "[[:alnum:]_]([_-]*[[:alnum:]_])*"
+#define SST_ALLOWED_METHODS_REGEX_PATTERN \
+    "^[[:blank:]]*(" METHOD_REGEX COLON_REGEX ")*" METHOD_REGEX "[[:blank:]]*$"
+
+static Regex sst_allowed_methods_regex;
+
+/* We allow custom sst scripts, so data can be anything.
+
+   We could create and maintain the list of forbidden
+   characters and the ways they could be used to inject the command to
+   the OS. However this approach seems to be too error prone.
+   Instead of this we will just allow alpha-num + a few special characters
+   (colon, slash, dot, underscore, square brackets).
+
+   It is the same as ^[\w:\/\.\[\]]+$    */
+
+   /* For some reason regex engine is not able to handle the following pattern
+   allowing square brackets as well:
+   "^[[:alnum:]:\\/\\._\\[\\]]+$"
+   We will substitute square brackets with lt/gt for validation */
+static const char *sst_method_allowed_chars_regex_pattern=
+    "^[[:alnum:]:\\/\\._<>]+$";
```

Comment:
Looks like this doesn't handle the case mentioned in the PXC-3456. (https://regexr.com/5e4hp). 
I think we should even allow `-` character as well.